crates/rustfix/src/replace.rs

```diff
@@ -97,96 +152,41 @@ impl Data {
             return Err(Error::DataLengthExceeded(range, self.original.len()));
         }
 
-        let insert_only = range.start == range.end;
-
-        // Since we error out when replacing an already replaced chunk of data,
-        // we can take some shortcuts here. For example, there can be no
-        // overlapping replacements -- we _always_ split a chunk of 'initial'
-        // data into three[^empty] parts, and there can't ever be two 'initial'
-        // parts touching.
-        //
-        // [^empty]: Leading and trailing ones might be empty if we replace
-        // the whole chunk. As an optimization and without loss of generality we
-        // don't add empty parts.
-        let Some(index_of_part_to_split) = self
-            .parts
-            .iter()
-            .position(|p| !p.data.is_inserted() && p.start <= range.start && p.end >= range.end)
-        else {
-            tracing::debug!(
-                "no single slice covering {}..{}, current slices: {:?}",
-                range.start,
-                range.end,
-                self.parts,
-            );
-            return Err(Error::MaybeAlreadyReplaced(range));
-        };
-
-        let part_to_split = &self.parts[index_of_part_to_split];
-
-        // If this replacement matches exactly the part that we would
-        // otherwise split then we ignore this for now. This means that you
-        // can replace the exact same range with the exact same content
-        // multiple times and we'll process and allow it.
-        //
-        // This is currently done to alleviate issues like
-        // rust-lang/rust#51211 although this clause likely wants to be
-        // removed if that's fixed deeper in the compiler.
-        if part_to_split.start == range.start && part_to_split.end == range.end {
-            if let State::Replaced(ref replacement) = part_to_split.data {
-                if &**replacement == data {
-                    return Ok(());
-                }
-            }
-        }
-
-        if part_to_split.data != State::Initial {
-            return Err(Error::AlreadyReplaced);
-        }
-
-        let mut new_parts = Vec::with_capacity(self.parts.len() + 2);
+        // Keep sorted by start position, with pure inserts before replacements.
+        let ins_point = self.parts.partition_point(|span| {
+            span.range.start < range.start
+                || (span.range.start == range.start && span.range.end < range.end)
```

Comment:
Could you explain a bit why we need `span.range.end < range.end`?

Reply:
> with pure inserts before replacements.

I think got it :)

Reply:
Yeah, it's to keep inserts ahead, but maybe I should elaborate in the comment why that's important: it ensures the checks that come next are valid, particularly if there might be multiple inserts/replacements at the same point. A single insert and replacement can have the same start, but any others at that point need to return an error.

Reply:
Sounds good to me!